### ISSUE TITLE
Form validation is not updated when fields are added to a form

### ISSUE THREAD

**Issue by estephan** [marmelab/react-admin#4169]
**What you were expecting:**

When a field is dynamically added to a form and that field impacts validation, I expect validation to be rerun so that the correct errors (or no errors) are displayed on the form.

**What happened instead:**
The form is unable to be saved as it still "has errors" that actually shouldn't exist.

**Steps to reproduce:**

1. Create a form with two `BooleanInputs` `a1` and `a2`.
2. Add a `defaultValue` to `a2` of `true`.
3. Make `a2` only display on the form when the value of `a1` is `true`.
4. Add validation to the form that requires the value of `a2` to be `true`
5. If on initial render of the form you click "save" a notification will appear saying "The form is not valid. Please check for errors", which is expected. however if you toggle `a1` and click "save" you still get the same notification (even though the values of the form are valid).

**Related code:**
I've modified the "Create Post" form to illustrate the bug.
https://codesandbox.io/s/agitated-feather-r98vi

```
<Create {...props}>
      <SimpleForm
        validate={values => {
          const errors = {};
          if (!values.a2) {
            errors.a2 = "Should be true";
          }
          return errors;
        }}
      >
        <BooleanInput source="a1" />
        <FormSpy subscription={{ values: true }}>
          {({ values }) =>
            values.a1 ? (
              <BooleanInput source="a2" defaultValue={true} />
            ) : null
          }
        </FormSpy>
      </SimpleForm>
    </Create>
```

**Other information:**

I have tested a similar thing using string inputs in `react-final-form` to see if the issue lies there. But it seems to work as expected.
https://codesandbox.io/s/react-final-form-field-arrays-jy77r

**Environment**

* React-admin version: 3.0.4
* Last version that did not exhibit the issue (if applicable): 2.x
* React version: 16.9.0
* Browser: Chrome (v 78.0.3904.108)

**Comment by fzaninotto** [comment]
I believe that's because the default subscription of the SimpleForm is:

```js

const defaultSubscription = {
    submitting: true,
    pristine: true,
    valid: true,
    invalid: true,
};
```

Subscriptions are [not well documented on the react-final-form doc](https://final-form.org/docs/react-final-form/types/FormSpyProps#subscription), but they allow better form performance at the price of disabling rare features. I think you're in the case where you need a custom subscription.

It can only be done in the `next` branch (upcoming 3.1 release) by passing a `subscription` prop to the `SimpleForm` component. Something like that should do the trick:

```diff
const defaultSubscription = {
    submitting: true,
    pristine: true,
    valid: true,
    invalid: true,
+  dirtyFields: true,
};
```

Not sure how you can test this trick yet. I'm keeping this issue opened as we should probably document this.

**Comment by estephan** [comment]
So I tried the trick of adding `dirtyFields: true` to `defaultSubscription` which unfortunately didn't seem to solve it. I then removed the whole `subscription` prop from being passed in to `Form` in `TabbedForm.ts` and still no luck.

And then to _work around_ this issue I tried to pass in `validateOnBlur` as a prop to `TabbedForm`, which was better until I realised in RA3, clicking the "Save" button doesn't blur the last input the user was focused on. This led to forms being able to be submitted with failing validation. 
Logged a separate bug [here](https://github.com/marmelab/react-admin/issues/4181).
